compiler/src/iree/compiler/Dialect/LinalgExt/Transforms/SplitReduction.cpp

```diff
@@ -389,4 +390,352 @@ splitReduction(RewriterBase &rewriter, LinalgExt::TopkOp topkOp,
+    // necessary.
+    return rewriter.notifyMatchFailure(
+        genericOp, "innerParallel split is not supported yet");
+  }
+
+  int64_t ratio = control.ratio;
+  unsigned insertSplitIndex = control.index;
+  unsigned insertSplitDimension = control.index;
+  if (ratio <= 1) {
+    return rewriter.notifyMatchFailure(
+        genericOp, "split ratio needs to be greater than 1");
+  }
+
+  SmallVector<unsigned> dims;
+  genericOp.getReductionDims(dims);
+  unsigned reductionDim = dims[0];
+  if (control.innerParallel) {
+    insertSplitDimension = reductionDim + 1;
+  }
+
+  SmallVector<int64_t, 4> loopRanges = genericOp.getStaticLoopRanges();
+  int64_t reductionDimSize = loopRanges[reductionDim];
+
+  if (ShapedType::isDynamic(reductionDimSize) ||
+      reductionDimSize % ratio != 0) {
+    return rewriter.notifyMatchFailure(
+        genericOp, "Reduction dimension not divisible by split ratio");
+  }
+
+  if (insertSplitIndex >
+      genericOp.getShape(genericOp.getDpsInitOperand(0)).size()) {
+    return rewriter.notifyMatchFailure(genericOp,
+                                       "Insert dimension position too large "
+                                       "compared to intermediate tensor size");
+  }
+
+  ArgmaxCombinerOps combinerOps = collectArgmaxCombinerOps(genericOp);
+  Operation *reductionOp = combinerOps.maxOp;
+
+  std::optional<TypedAttr> identity = arith::getNeutralElement(reductionOp);
+  if (!identity.has_value())
+    return rewriter.notifyMatchFailure(
+        genericOp, "Unknown identity value for the reduction");
+
+  SmallVector<Value> newInputs;
+  for (OpOperand *operand : genericOp.getDpsInputOperands()) {
+    AffineMap map = genericOp.getMatchingIndexingMap(operand);
+    ArrayRef<int64_t> oldShape = genericOp.getShape(operand);
+    Type elementType =
+        cast<RankedTensorType>(operand->get().getType()).getElementType();
+
+    SmallVector<int64_t> expandedShape;
+    for (unsigned idx : llvm::seq<unsigned>(0u, map.getNumResults())) {
+      unsigned dim = map.getDimPosition(idx);
+      if (dim == reductionDim) {
+        int64_t orig = oldShape[idx];
+        int64_t outer = orig / ratio;
+        if (control.innerParallel) {
+          expandedShape.push_back(ratio); // reduction.
+          expandedShape.push_back(outer); // parallel.
+        } else {
+          expandedShape.push_back(outer); // parallel.
+          expandedShape.push_back(ratio); // reduction.
+        }
+      } else {
+        expandedShape.push_back(oldShape[idx]);
+      }
+    }
+
+    RankedTensorType expandedType =
+        RankedTensorType::get(expandedShape, elementType);
+    Value expanded = expandValue(rewriter, loc, operand->get(), expandedType);
+    newInputs.push_back(expanded);
+  }
+
+  //  The total number of output elements along this new dimension is
+  //  reductionDimSize / ratio.
+  int64_t outputDimsize = reductionDimSize / ratio;
```

Comment:
style nit: I think you accidentally add one more leading whitespace in the comment. :)
nit 2: can we rename it to `outputDimSize` for consistency?

```suggestion
  // The total number of output elements along this new dimension is
  // reductionDimSize / ratio.
  int64_t outputDimSize = reductionDimSize / ratio;
```